src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1003,15 +1000,15 @@ private void onDecodePublish(
                 final MessageConsumer newTarget = router.supplyReceiver(newInitialId);
 
                 final int topicKey = topicKey(topicName);
-                MqttServer.MqttPublishStream publishStream = publishers.computeIfAbsent(topicKey, s ->
+                MqttServerStream stream = streams.computeIfAbsent(topicKey, s ->
                 {
-                    final MqttPublishStream newPublishStream = new MqttPublishStream(newTarget,
-                                                                     newRouteId, newInitialId, newReplyId, 0, topicName);
-                    newPublishStream.doApplicationBegin(traceId, authorization, affinity);
-                    return newPublishStream;
+                    final MqttServerStream newStream = new MqttServerStream(newTarget, newRouteId, newInitialId, newReplyId,
+                                                        0, topicName, null, MqttRole.SENDER);
+                    newStream.doApplicationBegin(traceId, authorization, affinity, topicName, 0);
+                    return newStream;
                 });
-                publishStream.doApplicationData(traceId, authorization, payload, dataEx);
-                correlations.put(newReplyId, publishStream);    // TODO: do we need to clean up correlations onAbort()?
+                stream.doApplicationData(traceId, authorization, MqttRole.SENDER, payload, dataEx);
+                correlations.put(newReplyId, stream);    // TODO: do we need to clean up correlations onAbort()?
```

Comment:
Yes, correlations need to be tidied up for each application stream.
This can be handled when the application stream is asked to cleanup.